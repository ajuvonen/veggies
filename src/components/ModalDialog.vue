<script setup lang="ts">
import {Dialog, DialogPanel, DialogTitle, TransitionRoot} from '@headlessui/vue';

const model = defineModel<boolean>({required: true});
defineProps<{
  title: string;
}>();
</script>
<template>
  <TransitionRoot
    :show="model"
    as="template"
    enter="duration-200 ease-out"
    enter-from="opacity-0"
    enter-to="opacity-100"
    leave="duration-200 ease-in"
    leave-from="opacity-100"
    leave-to="opacity-0"
  >
    <Dialog static class="relative z-10" @close="model = false">
      <!-- The backdrop, rendered as a fixed sibling to the panel container -->
      <div class="modal-dialog__backdrop" aria-hidden="true" />
      <!-- Full-screen container to center the panel -->
      <div class="modal-dialog__container">
        <DialogPanel class="modal-dialog outline-override">
          <ButtonComponent
            :aria-label="$t('general.close')"
            class="modal-dialog__close-button"
            variant="text"
            icon="close"
            data-test-id="close-button"
            @click="model = false"
          />
          <DialogTitle as="h2" class="modal-dialog__title">{{ title }}</DialogTitle>
          <div class="modal-dialog__content outline-override">
            <slot name="content"></slot>
          </div>
          <div class="modal-dialog__buttons outline-override">
            <slot name="buttons"> </slot>
          </div>
        </DialogPanel>
      </div>
    </Dialog>
  </TransitionRoot>
</template>
<style lang="scss" scoped>
.modal-dialog__backdrop {
  @apply fixed inset-0;
  @apply bg-black/30;
}

.modal-dialog__container {
  @apply fixed inset-0 w-screen p-4;
  @apply flex items-center justify-center;
}

.modal-dialog {
  @apply w-full max-w-xl max-h-full rounded-md p-4 shadow-xl relative;
  @apply flex flex-col gap-4;
  @apply bg-slate-50;
}

.modal-dialog__title {
  @apply text-lg leading-6;
}

.modal-dialog__content {
  @apply has-scroll;
  @apply flex flex-col gap-4;
  scrollbar-color: initial;
}

.modal-dialog__title,
.modal-dialog__content {
  @apply text-slate-900;
}

.modal-dialog__close-button {
  @apply absolute right-4 top-4;
  @apply fill-slate-700;
}

.modal-dialog__buttons {
  @apply flex-container justify-end;
}
</style>
