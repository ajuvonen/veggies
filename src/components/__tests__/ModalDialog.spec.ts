import {describe, it, expect} from 'vitest';
import {mount, type ComponentMountingOptions} from '@vue/test-utils';
import ModalDialog from '@/components/ModalDialog.vue';
import DialogStub from './DialogStub.vue';

const mounter = (options?: Partial<ComponentMountingOptions<typeof ModalDialog>>) =>
  mount(ModalDialog, {
    ...options,
    global: {
      stubs: {
        Dialog: DialogStub,
        DialogPanel: {
          template: '<div><slot /></div>',
        },
        DialogTitle: true,
      },
    },
  });

describe('ModalDialog', () => {
  it('renders closed', () => {
    const wrapper = mount(ModalDialog, {
      props: {
        title: 'Test dialog',
        modelValue: false,
      },
    });
    expect(wrapper).toBeTruthy();
    expect(wrapper.findByTestId('dialog').exists()).toBe(false);
  });

  it('shows dialog', () => {
    const wrapper = mounter({
      props: {
        title: 'Test dialog',
        modelValue: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findByTestId('dialog').exists()).toBe(true);
    expect(wrapper.findByTestId('dialog-close-button').exists()).toBe(true);
  });

  it('renders content', () => {
    const wrapper = mounter({
      props: {
        title: 'Test dialog',
        modelValue: true,
      },
      slots: {
        content: '<p>Test content</p>',
        buttons: '<p>Test buttons</p>',
      },
    });
    expect(wrapper.findByText('p', 'Test content').exists()).toBe(true);
    expect(wrapper.findByText('p', 'Test buttons').exists()).toBe(true);
    expect(wrapper.findByTestId('dialog-close-button').exists()).toBe(false);
  });

  it('closes dialog on button click', async () => {
    const wrapper = mounter({
      props: {
        title: 'Test dialog',
        modelValue: true,
      },
    });
    await wrapper.findByTestId('dialog-close-button').trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([[false]]);
  });
});
