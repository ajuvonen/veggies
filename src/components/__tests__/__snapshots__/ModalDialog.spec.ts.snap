// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalDialog > shows dialog 1`] = `
"<div data-v-0f2dec55="" class="relative z-10">
  <!-- The backdrop, rendered as a fixed sibling to the panel container -->
  <div data-v-0f2dec55="" class="modal-dialog__backdrop" aria-hidden="true"></div><!-- Full-screen container to center the panel -->
  <div data-v-0f2dec55="" class="modal-dialog__container">
    <div data-v-0f2dec55="" class="modal-dialog outline-override"><button data-v-24e38f3d="" data-v-0f2dec55="" class="button button--text modal-dialog__close-button" aria-label="Close" data-test-id="close-button"><svg data-v-24e38f3d="" viewBox="0 0 24 24" aria-hidden="true" style="width: 1.25rem; height: 1.25rem;">
          <path d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z" class="transition-all"></path>
        </svg></button>
      <dialog-title-stub data-v-0f2dec55="" as="h2" class="modal-dialog__title"></dialog-title-stub>
      <div data-v-0f2dec55="" class="modal-dialog__content outline-override"></div>
      <div data-v-0f2dec55="" class="modal-dialog__buttons outline-override"></div>
    </div>
  </div>
</div>"
`;
